backend/shade/shade.go

```diff
@@ -0,0 +1,1086 @@
+			return false, err
+		}
+		return res.StatusCode == http.StatusTooManyRequests, nil
+	})
+
+	if err != nil {
+		fs.Debugf(o.fs, "Download request failed: %v", err)
+		return nil, fmt.Errorf("download request failed: %w", err)
+	}
+	if res == nil {
+		return nil, fmt.Errorf("no response received from initial request")
+	}
+
+	// Handle response based on status code
+	switch res.StatusCode {
+	case http.StatusOK:
+		fs.Debugf(o.fs, "Received file directly from ShadeFS")
+		return res.Body, nil
+
+	case http.StatusTemporaryRedirect:
+		// Read the presigned URL from the body
+		bodyBytes, err := io.ReadAll(res.Body)
+		fs.CheckClose(res.Body, &err) // Close body after reading
+		if err != nil {
+			fs.Debugf(o.fs, "Failed to read redirect body: %v", err)
+			return nil, fmt.Errorf("failed to read redirect body: %w", err)
+		}
+
+		presignedURL := strings.TrimSpace(string(bodyBytes))
+		o.original = presignedURL //Save for later for hashing
+
+		client := rest.NewClient(fshttp.NewClient(ctx)).SetRoot(presignedURL)
+		var downloadRes *http.Response
+		opts := rest.Opts{
+			Method:  "GET",
+			Path:    "",
+			Options: options,
+		}
+
+		err = o.fs.pacer.Call(func() (bool, error) {
+			fs.Debugf(o.fs, "Downloading file: %s", presignedURL)
+			downloadRes, err = client.Call(ctx, &opts)
+			if err != nil {
+				fs.Debugf(o.fs, "Failed to fetch presigned URL: %v", err)
+				return false, err
+			}
+			if downloadRes == nil {
+				return false, fmt.Errorf("failed to fetch presigned URL")
+			}
+			return downloadRes.StatusCode == http.StatusTooManyRequests, nil
+		})
+
+		if err != nil {
+			return nil, fmt.Errorf("presigned URL request failed: %w", err)
+		}
+		if downloadRes == nil {
+			return nil, fmt.Errorf("no response received from presigned URL request")
+		}
+
+		if downloadRes.StatusCode != http.StatusOK && downloadRes.StatusCode != http.StatusPartialContent {
+			body, _ := io.ReadAll(downloadRes.Body)
+			fs.CheckClose(downloadRes.Body, &err)
+			return nil, fmt.Errorf("presigned URL request failed with status %d: %q", downloadRes.StatusCode, string(body))
+		}
+
+		return downloadRes.Body, nil
+
+	default:
+		body, _ := io.ReadAll(res.Body)
+		fs.CheckClose(res.Body, &err)
+		fs.Debugf(o.fs, "Unexpected status code from ShadeFS: %d, body: %q", res.StatusCode, string(body))
+		return nil, fmt.Errorf("download failed with status %d: %q", res.StatusCode, string(body))
+	}
+}
+
+// Update in to the object with the modTime given of the given size
+//
+// When called from outside an Fs by rclone, src.Size() will always be >= 0.
+// But for unknown-sized objects (indicated by src.Size() == -1), Upload should either
+// return an error or update the object properly (rather than e.g. calling panic).
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	fs.Debugf(o.fs, "Uploading file: %s", o.remote)
+
+	if src.Size() < 0 {
+		//Indeterminate sizes not supported
+		return fs.ErrorCantMove
```

Comment:
This is the wrong error. It shouldn't happen in normal use as the upper layers will ensure it doesn't get called but that error will definitely confuse the users! I'd probably use `errors.New("file size unknown")`